fix(neuron.connection.http): fix error case so return value matches t…

krisalyssa opened the pull request:
…ypespec

The definition of the `JSONParseError` struct says that the `response` element is a
`Neuron.Response.t()`, but the code was actually returning a tuple (`{:error,
Neuron.Response.t()}`). With this change the type of `response` matches the typespec.

BREAKING CHANGE: If your code expects the `response` field of `JSONParseError` to be a tuple, it
will break.

fix #59

Commit:
fix(neuron.connection.http): fix error case so return value matches typespec

The definition of the `JSONParseError` struct says that the `response` element is a
`Neuron.Response.t()`, but the code was actually returning a tuple (`{:error,
Neuron.Response.t()}`). With this change the type of `response` matches the typespec.

BREAKING CHANGE: If your code expects the `response` field of `JSONParseError` to be a tuple, it
will break.

fix #59

uesteibar commented:
Hi there! thanks for taking care of this ❤️ however there is some tests failing that we should fix first, you can see it on https://travis-ci.org/github/uesteibar/neuron/builds/685800174.

Not sure why it's not showing in here, I'll investigate

krisalyssa commented:
I'm finally circling back to check on these failing tests. The first thing I notice is that TravisCI is building using older versions of Elixir and Erlang than I did when developing, and it looks like something's trying to use a function that's not available in the older Erlang. I'm installing OTP 19.0 and Elixir 1.6.6 to see if and how I need to adjust my PR.

krisalyssa commented:
@uesteibar I was unable to get OTP 19.0 to install successfully on macOS 10.15, so I can't recreate the error that TravisCI is reporting. So I'm making educated guesses based on what I can see in TravisCI, and editing the code in the PR to what I hope will be more OTP 19-friendly. There may be several code pushes until I get it right, since the only way I can tell is to have TravisCI run the tests here.

Commit:
fix(neuron.connection.http): modify tests to work with OTP 19

The test added in the previous commit for this PR was developed against OTP, and apparently contains
code which isn't compatible with OTP 19 (with which TravisCI runs the tests).

fix #59

krisalyssa commented:
That push appears to have fixed the tests.

uesteibar commented on lib/neuron/connection/http.ex:
```diff
@@ -40,33 +40,33 @@ defmodule Neuron.Connection.Http do
   defp handle({:error, _} = response, _, _), do: response
 
-  defp build_response(%{status_code: 200} = response) do
+  defp build_response(response) do
```
Can we change the order and put `build_response` _after_ `build_response_tuple`? The former is used by the later so that is (imo) easier to read.

krisalyssa replied:
Done.

uesteibar reviewed (commented):
Thanks for taking care of this! just a tiny thing, after that I'll merge and release 👏 ❤️

Commit:
fix(neuron.connection.http): rearrange methods

Move `build_response/1` to after `build_response_tuple/1` since the latter calls the former

fix #59

uesteibar commented:
Published as `v5.0.0` 🚀 thank you!

krisalyssa commented:
You're very welcome! `neuron` has been useful to us, so I'm happy to give something back.
